### PR TITLE
Adding metaMDBG into YEAT

### LINKED ISSUE #57: Adding `metaMDBG` into YEAT
`metaMDBG` is a phenomenal tool for assembling metagenomic data. During my testing, I found that I could construct small plasmids where other popular long-read assemblers could not.

One of the biggest roadblocks to adding `metaMDBG` into YEAT's workflow is that it cannot be installed in the same Conda environment as YEAT. This is because of the incompatible packages of `openssl` and `wfmash` (see the following comment below).

To fix this, we'll need to have `metaMDBG` have its own conda environment for YEAT to activate during the snakemake workflow.

Another problem that arises with installing `metaMDBG` is that `wfmash` can only be installed in Linux environments. As a result, `metaMDBG` cannot be used on macOSX.

Even though the required OS environment of `metaMDBG` is a bit concerning, the situation has also been helpful in a way. Because of the package conflicts and the required OS, we can separate `metaMDBG` out into its own environment and do some checks to see if it exists.

### PR DESCRIPTION
**Pull request by danejo3**
The purpose of this PR is to resolve #57 .

In this PR, the assembly algorithm `metaMDBG` was added into YEAT's workflow; however, this algorithm is only available for Linux OS users. If the user tries to include `metaMDBG` in their config file and their system is not a linux OS, YEAT will stop and notify users to look at their config file.

In addition to adding `metaMDBG` into the workflow, I have spent sometime cleaning up the `makefile` and pytests.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -7,40 +7,48 @@ PYFILES = $(shell ls yeat/*.py yeat/*/*.py)
...
-## test:        run automated test suite
+## test:        run only short-running automated tests
 test:
-	pytest --cov=yeat -m 'not long and not bandage and not hifi and not nano'
+	pytest --cov=yeat -m short
 
 ## testlong:    run only long-running automated tests
 testlong:
-	pytest --cov=yeat -m 'long and not bandage and not hifi'
+	pytest --cov=yeat -m long
```
Comment (danejo3) [inline Makefile:16]: Added a bunch of pytest marks to simplify the input for `pytest -m`. Instead of having a long conditional `not` and `not` and so forth, we can just condense it down to one pytest mark.

File: Makefile [hunk 2]
```diff
@@ -7,40 +7,48 @@ PYFILES = $(shell ls yeat/*.py yeat/*/*.py)
...
-	pytest --cov=yeat -m 'nano and not long and not bandage'
+	pytest --cov=yeat -m nano
+
+## testbandage: run only Bandage required automated tests
+testbandage:
+	pytest --cov=yeat -m bandage
+
+## testlinux:   run only Linux required automated tests
+testlinux:
+	pytest --cov=yeat -m linux
```
Comment (danejo3) [inline Makefile:36]: New pytest mark for running Linux only tests. Currently, there is only one linux test in the test suite.

File: README.md [hunk 3]
```diff
@@ -20,16 +20,51 @@ Once the binary file has been obtained, users will need to update their environm
...
 Example for Ubuntu:
 ```
-export PATH={BANDAGE_DIR}:$PATH
+export PATH=~/projects/Bandage:$PATH
 ```
 
 Example for MacOS:
 ```
-export PATH={BANDAGE_DIR}/Bandage.app/Contents/MacOS:$PATH
+export PATH=~/projects/Bandage/Bandage.app/Contents/MacOS:$PATH
```
Comment (danejo3) [inline README.md:28]: As I was installing YEAT on my WSL2 machine and following the README, I added the curly braces for Bandage into my PATH. It took me a few minutes to figure out what I did wrong. I wouldn't be surprised if others will make the same mistake as me. Maybe showing an example path would be more beneficial than "copying-and-replacing" the `{BANDAGE_DIR}`?

File: README.md [hunk 4]
```diff
@@ -20,16 +20,51 @@ Once the binary file has been obtained, users will need to update their environm
...
+
+# Activate metaMDBG environment
+conda activate metaMDBG
+
+# Compile the software
+mkdir build
+cd build
+cmake ..
+make -j 3
+```
```
Comment (danejo3) [inline README.md:59]: Instructions to install `metaMDBG`. These were taken directly from their GitHub page: https://github.com/GaetanBenoitDev/metaMDBG#installation.

File: README.md [hunk 5]
```diff
@@ -20,16 +20,51 @@ Once the binary file has been obtained, users will need to update their environm
...
+
+After successful installation, an executable named metaMDBG will appear in ./build/bin.
+
+#### Link the binary to the environment's bin
+```
+ln -s ~/projects/metaMDBG/build/bin/metaMDBG /home/danejo/anaconda3/envs/metaMDBG/bin/metaMDBG
+```
```
Comment (danejo3) [inline README.md:66]: In order to run `metaMDBG` from anywhere, users will either have to update their PATH or "copy-paste"/symlink the binary file into their PATH. In this example, I have the binary file symlinked to my `bin`.

File: yeat/workflows/snakefiles/Pacbio [hunk 6]
```diff
@@ -124,3 +124,22 @@ rule unicycler:
...
+        "metaMDBG"
+    threads: 128
+    params:
+        outdir=lambda wildcards: f"analysis/{wildcards.sample}/{wildcards.label}/metamdbg",
+        extra_args=lambda wildcards: config["extra_args"][wildcards.label]
+    shell:
+        """
+        metaMDBG asm {params.outdir} {input.reads} -t {threads} {params.extra_args}
+        gunzip {params.outdir}/contigs.fasta.gz
+        ln -s contigs.fasta {output.contigs}
```
Comment (danejo3) [inline yeat/workflows/snakefiles/Pacbio:144]: Interestingly, `metaMDBG`'s final output is a .gz file of the contigs. In this rule, we uncompress the contig file and then symlink it.

File: yeat/workflows/snakefiles/Bandage [hunk 7]
```diff
@@ -137,3 +137,17 @@ rule hifiasm_meta:
...
+        status="analysis/{sample}/{label}/metamdbg/bandage/.done"
+    input:
+        gfa="analysis/{sample}/{label}/metamdbg/tmp/assembly_graph.gfa"
+    params:
+        outdir=lambda wildcards: f"analysis/{wildcards.sample}/{wildcards.label}/metamdbg/bandage"
+    shell:
+        """
+        Bandage image {input.gfa} {params.outdir}/assembly_graph.jpg
+        touch {output.status}
+        """
```
Comment (danejo3) [inline yeat/workflows/snakefiles/Bandage:153]: See https://github.com/bioforensics/yeat/issues/57#issuecomment-1710681608

File: yeat/workflows/snakefiles/Pacbio [hunk 8]
```diff
@@ -124,3 +124,22 @@ rule unicycler:
...
         """
+
+
+rule metamdbg:
+    output:
+        contigs="analysis/{sample}/{label}/metamdbg/{sample}_contigs.fasta"
+    input:
+        reads="seq/input/{sample}/combined-reads.fq.gz"
+    conda:
+        "metaMDBG"
```
Comment (danejo3) [inline yeat/workflows/snakefiles/Pacbio:135]: See #57 . 

TLDR: `metaMDBG` packages are not compatible with YEAT's environmental packages. Activating `metaMDBG` environment whenever the rule `metamdbg` is called.

**Review by danejo3 — COMMENTED** [review]
Comments on current PR.

File: yeat/cli/config.py [hunk 9]
```diff
@@ -182,17 +183,22 @@ def to_string(self):
...
         if self.algorithm == "canu":
             self.check_canu_required_params()
 
+    def check_valid_algorithm(self, data):
+        if data["algorithm"] not in ALGORITHMS:
+            message = f"Unsupported assembly algorithm '{data['algorithm']}'"
+            raise AssemblyConfigurationError(message)
+        if data["algorithm"] == "metamdbg" and platform not in ["linux", "linux2"]:
+            message = f"Assembly algorithm 'metaMDBG' can only run on Linux OS"
+            raise AssemblyConfigurationError(message)
```
Comment (danejo3) [inline yeat/cli/config.py:200]: Platform check statement. If the user adds the algorithm `metamdbg` into the config but the OS platform is not `linux`, an `AssemblyConfigurationError` will popup.

**Review by danejo3 — COMMENTED** [review]
OS platform check

**Comment by danejo3** [comment]
This PR is ready for review! The test suite takes a while to run!

File: yeat/tests/test_cli.py [hunk 10]
```diff
@@ -14,6 +14,8 @@
 from yeat.cli import InitAction, illumina
 from yeat.tests import data_file
 
+pytestmark = pytest.mark.short
+
```
Comment (standage) [inline yeat/tests/test_cli.py:18]: Curious: what is the purpose of this statement? Does it apply the mark to all tests in the file?

**Review by standage — APPROVED** [review]
The code looks good, and the tests passed for me. Thanks!

File: yeat/tests/test_cli.py (hunk 10 above)
Comment (danejo3) [reply yeat/tests/test_cli.py:18]: > Does it apply the mark to all tests in the file?

Yes! It's a quick way to apply the mark decorator on all the functions in that file.
